src/network/networkd-wiphy.c

```diff
@@ -80,6 +84,188 @@ int wiphy_get_by_name(Manager *manager, const char *name, Wiphy **ret) {
+         * To support that, we need to investigate what happens when the _phy_ is renamed. */
+
+        if (ret)
+                *ret = w;
+
+        return 0;
+}
+
+static int wiphy_get_rfkill_device(Wiphy *w, sd_device **ret) {
+        _cleanup_(sd_device_enumerator_unrefp) sd_device_enumerator *e = NULL;
+        sd_device *rfkill;
+        const char *s;
+        int r;
+
+        assert(w);
+        assert(w->name);
+
+        if (w->rfkill) {
+                if (ret)
+                        *ret = w->rfkill;
+
+                return 0;
+        }
+
+        if (!w->dev) {
+                r = sd_device_new_from_subsystem_sysname(&w->dev, "ieee80211", w->name);
+                if (r < 0)
+                        return r;
+        }
+
+        r = sd_device_get_syspath(w->dev, &s);
+        if (r < 0)
+                return r;
+
+        r = sd_device_enumerator_new(&e);
+        if (r < 0)
+                return r;
+
+        r = sd_device_enumerator_allow_uninitialized(e);
+        if (r < 0)
+                return r;
+
+        r = sd_device_enumerator_add_match_subsystem(e, "rfkill", true);
+        if (r < 0)
+                return r;
+
+        r = sd_device_enumerator_add_match_parent(e, w->dev);
+        if (r < 0)
+                return r;
+
+        rfkill = sd_device_enumerator_get_device_first(e);
+        if (!rfkill)
+                return log_wiphy_debug_errno(w, SYNTHETIC_ERRNO(ENODEV), "No rfkill device found.");
+
+        if (sd_device_enumerator_get_device_next(e))
+                return log_wiphy_debug_errno(w, SYNTHETIC_ERRNO(EEXIST), "Multiple rfkill devices found.");
+
+        w->rfkill = sd_device_ref(rfkill);
+        log_wiphy_debug(w, "rfkill device found: %s", sd_device_get_syspath(rfkill, &s) >= 0 ? s : "n/a");
+
+        if (ret)
+                *ret = w->rfkill;
+
+        return 0;
+}
+
+static int wiphy_get_rfkill_state(Wiphy *w) {
+        sd_device *dev;
+        int r;
+
+        assert(w);
+
+        r = wiphy_get_rfkill_device(w, &dev);
+        if (r < 0)
+                return r;
+
+        /* The previous values may be outdated. Let's clear cache and re-read the values. */
+        device_clear_sysattr_cache(dev);
+
+        r = device_get_sysattr_bool(dev, "soft");
+        if (r < 0 && r != -ENOENT)
+                return r;
+        if (r > 0)
+                return RFKILL_SOFT;
+
+        r = device_get_sysattr_bool(dev, "hard");
+        if (r < 0 && r != -ENOENT)
+                return r;
+        if (r > 0)
+                return RFKILL_HARD;
+
+        return RFKILL_UNBLOCKED;
+}
+
+int link_rfkilled(Link *link) {
+        Wiphy *w;
+        int r;
+
+        assert(link);
+
+        r = link_get_wiphy(link, &w);
+        if (IN_SET(r, -EOPNOTSUPP, -ENODEV))
+                return false; /* Typically, non-wifi interface. */
+        if (r < 0)
+                return log_link_debug_errno(link, r, "Could not get phy: %m");
```

Comment:
I would split this part off into `link_rfkilled()` and move the other part of this function to a new function `wiphy_rfkilled()`. `link_rfkilled()` and `link_get_wiphy()` can then remain in network-link.c. `link_rfkilled()` will then just call `link_get_wiphy()` followed by `wiphy_rfkilled()`